scheduler/src/cook/kubernetes/compute_cluster.clj

```diff
@@ -0,0 +1,385 @@
+                                   )
+          node-name->node (pc/map-from-vals (fn [^V1Node node]
+                                              (-> node .getMetadata .getName))
+                                            (.getItems current-nodes))]
+      (reset! current-nodes-atom node-name->node)
+      (let [watch (WatchHelper/createNodeWatch api-client (-> current-nodes .getMetadata .getResourceVersion))]
+        (.submit kubernetes-executor ^Callable
+          (fn []
+            (try
+              (handle-watch-updates current-nodes-atom watch (fn [n] (-> n .getMetadata .getName)) nil)
+              (catch Exception e
+                (log/warn e "Error during node watch")
+                (initialize-node-watch api-client))
+              (finally
+                (.close watch))))))))
+
+  (defn get-nodes
+    []
+    @current-nodes-atom))
+
+(defn to-double
+  [^Quantity q]
+  (-> q .getNumber .doubleValue))
+
+(defn convert-resource-map
+  [m]
+  {:mem (if (get m "memory")
+          (-> m (get "memory") to-double (/ (* 1024 1024)))
+          0.0)
+   :cpus (if (get m "cpu")
+           (-> m (get "cpu") to-double)
+           0.0)})
+
+(defn get-capacity
+  [node-name->node]
+  (pc/map-vals (fn [^V1Node node]
+                 (-> node .getStatus .getCapacity convert-resource-map))
+               node-name->node))
+
+(defn get-consumption
+  [pod-name->pod]
+  (let [node-name->pods (group-by (fn [^V1Pod p] (-> p .getSpec .getNodeName))
+                                  (vals pod-name->pod))
+        node-name->requests (pc/map-vals (fn [pods]
+                                           (->> pods
+                                                (map (fn [^V1Pod pod]
+                                                       (let [containers (-> pod .getSpec .getContainers)
+                                                             container-requests (map (fn [^V1Container c]
+                                                                                       (-> c
+                                                                                           .getResources
+                                                                                           .getRequests
+                                                                                           convert-resource-map))
+                                                                                     containers)]
+                                                         (apply merge-with + container-requests))))
+                                                (apply merge-with +)))
+                                         node-name->pods)]
+    node-name->requests))
+
+(defn generate-offers
+  [node-name->node pod-name->pod compute-cluster]
+  (let [node-name->capacity (get-capacity node-name->node)
+        node-name->consumed (get-consumption pod-name->pod)
+        ; TODO(pschorf): Should also include recently launched jobs
```

Comment:
Is this TODO for this PR or for the future?

Reply:
These are all for the future, just wanted to leave some breadcrumbs for myself